actionview/lib/action_view/helpers/translation_helper.rb

```diff
@@ -119,7 +119,7 @@ def localize(object, **options)
       alias :l :localize
 
       private
-        MISSING_TRANSLATION = Object.new
+        MISSING_TRANSLATION = -1327903252
```

Comment:
Interesting idea!  Though I am somewhat wary of choosing magic numbers.  `Class.new` also seems to work, but I'm not sure if it works with all cache stores (but maybe that doesn't matter, given the use case).

Another possibility would be to choose a sufficiently unlikely string (e.g. a string of unprintable characters), and change [`equal?`](https://github.com/rails/rails/blob/3cbc7279fbbe3d941f1b0aac4bc13c1620c3c4eb/actionview/lib/action_view/helpers/translation_helper.rb#L93) to `==`.

But I don't know how any of these options affect performance both when using `I18n::Backend::Cache` and not.  And I'm not sure if performance is even a practical concern when using `I18n::Backend::Cache` — see https://github.com/rails/rails/issues/45571#issuecomment-1181930469.

Reply:
> at least in redis/memcached, where objects are serialized/deserialized.

Objects are [serialized](https://github.com/rails/rails/blob/05ae3faaf3779d53935ab983b75f2dc7c6dfb7fa/activesupport/lib/active_support/cache/memory_store.rb#L188) for `MemoryStore` too, hence the current issue. :wink:  But yes, the implementation may be different.

> I am not sure this is better than just the number.

Generally speaking, using a random anything concerns me because nothing prevents someone else from "randomly" using that thing too.  Another concern, though, is that we'd still be relying on `equal?` to return true for a deserialized object.  For MRI, that's true for smaller numbers (like the one you've chosen), but do you know if that is part of the Ruby spec?

If not, I think using `==` would be safer.  And if we're using `==`, I think using an intentional string like `"\0\a"` or `"\aMISSING"` or similar makes sense.  Although if benchmarks show a significant difference, then I am open to using a number.  But I would feel better about using a larger number (or, in the negative, smaller).

Reply:
> I haven't measured, but for sure there will be only a minor difference, if any.

`I18n.translate` can have a [surprising performance difference](https://github.com/rails/rails/pull/44299#discussion_r798781173) depending on the type of its arguments.  I would encourage you to benchmark this. :wink: 

Additionally, whichever constant we choose, would mind also applying it [here](https://github.com/rails/rails/blob/4f818c7202ee26b8d33f8817c6064ba94e386596/activemodel/lib/active_model/naming.rb#L214) and [here](https://github.com/rails/rails/blob/4f818c7202ee26b8d33f8817c6064ba94e386596/activemodel/lib/active_model/translation.rb#L38)?

Reply:
I wanted to see how the results compared with `Object.new`, so I adapted my benchmark script from #44299:

<details>
  <summary><strong>Benchmark script</strong></summary>

  ```ruby
  require "benchmark/ips"

  class BaseModel
    extend ActiveModel::Translation
  end

  BlogPost = Class.new(BaseModel)

  NAMED_CLASS = Class.new

  [true, false].each do |missing|
    [false, true].each do |caching|
      I18n.backend = Class.new(I18n::Backend::Simple).new
      if caching
        I18n.backend.class.include(I18n::Backend::Cache)
        I18n.cache_store = ActiveSupport::Cache.lookup_store(:memory_store)
      end

      if !missing
        I18n.backend.store_translations(:en, activemodel: { models: { blog_post: "Blerg" } })
      end

      puts ""
      puts "--- missing: #{missing}, caching: #{caching} ".ljust(72, "-")
      puts ""

      [Object.new, NAMED_CLASS, "\aMISSING\a", 2**30, 2**60].each do |sentinel|
        ActiveModel::Name.send(:remove_const, :MISSING_TRANSLATION)
        ActiveModel::Name.const_set(:MISSING_TRANSLATION, sentinel)

        Benchmark.ips do |x|
          x.report(sentinel.inspect) { BlogPost.model_name.human }
        end
      end
    end
  end
  ```
</details>

The results I got appear to be different than yours:

```
--- missing: true, caching: false --------------------------------------

#<Object:0x00007f376...> 27.906k (± 1.3%) i/s -    140.550k in   5.037437s
         NAMED_CLASS     27.876k (± 0.8%) i/s -    140.913k in   5.055406s
       "\aMISSING\a"     13.512k (± 1.1%) i/s -     68.100k in   5.040470s
          1073741824     27.889k (± 1.0%) i/s -    139.850k in   5.015039s
 1152921504606846976     27.768k (± 1.2%) i/s -    139.850k in   5.037131s

--- missing: true, caching: true ---------------------------------------

#<Object:0x00007f376...> 21.390k (± 1.0%) i/s -    108.579k in   5.076757s
         NAMED_CLASS     23.524k (± 1.3%) i/s -    119.187k in   5.067600s
       "\aMISSING\a"     26.392k (± 0.9%) i/s -    132.300k in   5.013182s
          1073741824     27.862k (± 1.1%) i/s -    140.300k in   5.036137s
 1152921504606846976     27.827k (± 1.0%) i/s -    139.450k in   5.011886s

--- missing: false, caching: false -------------------------------------

#<Object:0x00007f2c4...> 18.859k (± 1.4%) i/s -     94.600k in   5.017336s
         NAMED_CLASS     18.952k (± 0.8%) i/s -     96.186k in   5.075477s
       "\aMISSING\a"     18.905k (± 1.4%) i/s -     96.084k in   5.083430s
          1073741824     18.827k (± 0.9%) i/s -     95.268k in   5.060580s
 1152921504606846976     18.734k (± 1.3%) i/s -     95.064k in   5.075230s

--- missing: false, caching: true --------------------------------------

#<Object:0x00007f376...> 24.845k (± 1.1%) i/s -    124.350k in   5.005617s
         NAMED_CLASS     26.684k (± 1.0%) i/s -    136.017k in   5.097946s
       "\aMISSING\a"     26.510k (± 1.0%) i/s -    134.844k in   5.087075s
          1073741824     26.715k (± 1.1%) i/s -    134.500k in   5.035266s
 1152921504606846976     26.728k (± 1.1%) i/s -    133.950k in   5.012120s
```

If you run my script, do you see similar results?  If so, what is causing the discrepancy with the results you posted?

In particular, I would like to point out that when "missing: true" (which is common enough for e.g. `model_name.human`), caching in almost every case appears to be slower than not caching while using `MISSING_TRANSLATION = Object.new`.  However, caching *does* appear to provide a significant speed-up when "missing: false".  Those are the kind of results I was looking for in https://github.com/rails/rails/issues/45571#issuecomment-1181930469.

On the plus side, it looks like there is a strong case for using a large integer for `MISSING_TRANSLATION`. :smiley:

Reply:
It was kinda obvious that the number should be faster than the string 😄 
I am seeing similar results with your benchmark. My benchmark was very simple and on the view context.
There is a very little probability someone will use a random 32-bits integer, but lets use a bigger integer, since it shows the same results.

And thank you for the patience and detailed comments ❤️! I hope this will be merged this time 😄